**INTRODUÇÃO**

Este é um experimento que tem como objetivo comparar duas bibliotecas de Deep Learning: TensorFlow(Keras) e PyTorch.

O objetivo é construir uma CNN(Convolutional Neural Network) nas duas bibliotecas, com a mesma arquitetura e os mesmos hiperparâmetros

O dataset escolhido foi o MNIST, este conjunto de dados é uma coleção de dígitos manuscritos, concebida para treinar e testar sistemas de classificação de imagens.

**METODOLOGIA**

A arquitetura da CNN (Convolutional Neural Network) é projetada para classificação de imagens em 10 classes. Abaixo está um resumo detalhado da arquitetura:

1.   Camada de Entrada

    *   A rede espera imagens de entrada com tamanho 28x28 pixels e 1 canal, o que sugere que as imagens são em escala de cinza (grayscale).


2.   Camadas Convolucionais e de Pooling:

    *   1ª Camada Convolucional:
      * Filtros: 28
      * Kernel Size: (3, 3)
      * Função de Ativação: ReLU
      * Esta camada aplica 28 filtros convolucionais com um kernel de tamanho 3x3 na imagem de entrada. A função de ativação ReLU é usada para introduzir não-linearidade.

    *   1ª Camada de Pooling:
      * Pool Size: (2, 2)
      * Esta camada reduz a dimensionalidade da imagem ao aplicar uma operação de pooling com uma janela de 2x2, o que reduz pela metade a resolução espacial.

    * 2ª Camada Convolucional:
      * Filtros: 64
      * Kernel Size: (3, 3)
      * Função de Ativação: ReLU
      * Outra camada convolucional que aumenta a profundidade para 64 filtros, mantendo o kernel 3x3.

    * 2ª Camada de Pooling:
      * Pool Size: (2, 2)
      * Outra operação de pooling para reduzir a resolução espacial, mantendo as características mais importantes.

    * 3ª Camada Convolucional:
      * Filtros: 64
      * Kernel Size: (3, 3)
      * Função de Ativação: ReLU
      * Mais uma camada convolucional com 64 filtros, que continua a extração de características mais profundas da imagem.
      
3. Camadas Densas:

    * Camada Flatten:
      * A saída das camadas convolucionais (que é uma matriz tridimensional) é achatada em um vetor unidimensional para ser usada como entrada para as camadas densas.

    * 1ª Camada Densa:
      * Neurônios: 64
      * Função de Ativação: ReLU
      * Esta camada totalmente conectada (fully connected) aprende combinações não-lineares das características extraídas pelas camadas convolucionais.

    * 2ª Camada Densa(Saída):
      * Neurônios: 10
      * Esta camada final produz 10 valores, um para cada classe, sem aplicar uma função de ativação porque a perda usada (SparseCategoricalCrossentropy) será aplicada diretamente aos logits.

4. Compilação do Modelo:

    * Otimizador: Adam com uma taxa de aprendizado de 0,001.
    * Função de Perda: SparseCategoricalCrossentropy (usada para tarefas de classificação onde as classes são representadas por rótulos inteiros).


**IMPLEMENTAÇÃO CNN TENSORFLOW**

In [ ]:
## Bibliotecas necessárias

import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras import models, layers
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np
import matplotlib.pyplot as plt
import time
import random

In [ ]:
seed = 42
tf.random.set_seed(seed)
np.random.seed(seed)
random.seed(seed)

In [ ]:
# Obtém o dataset
(x_train_original,y_train),(x_test_original, y_test) = mnist.load_data()

x_train = np.expand_dims(x_train_original, axis=-1)
x_test = np.expand_dims(x_test_original, axis=-1)

In [ ]:
x_train, x_test = x_train / 255.0, x_test / 255.0

In [ ]:
# Constrói o modelo CNN
model = models.Sequential()
model.add(layers.Conv2D(28, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# Compila o modelo
optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)

model.compile(optimizer=optimizer,
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])


In [ ]:
# Treina o modelo
start_time = time.time()
history = model.fit(x_train, y_train, epochs=5, batch_size=64,
                    validation_data=(x_test, y_test))
end_time = time.time()
keras_time = end_time - start_time



Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 58s 60ms/step - accuracy: 0.8732 - loss: 0.4227 - val_accuracy: 0.9802 - val_loss: 0.0593
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 53s 57ms/step - accuracy: 0.9834 - loss: 0.0524 - val_accuracy: 0.9862 - val_loss: 0.0467
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 50s 54ms/step - accuracy: 0.9889 - loss: 0.0356 - val_accuracy: 0.9820 - val_loss: 0.0562
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 82s 54ms/step - accuracy: 0.9921 - loss: 0.0271 - val_accuracy: 0.9786 - val_loss: 0.0760
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 83s 55ms/step - accuracy: 0.9938 - loss: 0.0219 - val_accuracy: 0.9867 - val_loss: 0.0440


**IMPLEMENTAÇÃO CNN PYTORCH**

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch.utils.data import DataLoader, TensorDataset
import torch.nn.functional as F
import matplotlib.pyplot as plt
import random

In [ ]:
seed = 42
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
np.random.seed(seed)
random.seed(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

In [ ]:
# Transformações para normalizar o dataset
transform = transforms.Compose([
    transforms.ToTensor()
])

In [ ]:
# Carregar o dataset MNIST
train_images = torch.tensor(x_train_original, dtype=torch.float32).unsqueeze(1) / 255.0
train_labels = torch.tensor(y_train, dtype=torch.long)
test_images = torch.tensor(x_test_original, dtype=torch.float32).unsqueeze(1) / 255.0
test_labels = torch.tensor(y_test, dtype=torch.long)

# Criar datasets e dataloaders
train_dataset = TensorDataset(train_images, train_labels)
test_dataset = TensorDataset(test_images, test_labels)

train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)

In [ ]:
# Criar DataLoader para facilitar o treinamento
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)

In [ ]:
# Definir a arquitetura da CNN
class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        # A camada Conv2D inicial de TensorFlow tinha 28 filtros
        self.conv1 = nn.Conv2d(1, 28, kernel_size=3, stride=1, padding=0)
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2, padding=0)
        self.conv2 = nn.Conv2d(28, 64, kernel_size=3, stride=1, padding=0)
        self.conv3 = nn.Conv2d(64, 64, kernel_size=3, stride=1, padding=0)
        self.fc1 = nn.Linear(64 * 3 * 3, 64)
        self.fc2 = nn.Linear(64, 10)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = self.pool(x)
        x = F.relu(self.conv2(x))
        x = self.pool(x)
        x = F.relu(self.conv3(x))
        x = x.view(-1, 64 * 3 * 3)  # Flatten
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x

In [ ]:
# Instanciar o modelo e definir a função de perda e otimizador
model_pytorch = CNN()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model_pytorch.parameters(), lr=0.001)

In [ ]:
# Treinamento do modelo
start_time = time.time()
for epoch in range(5):  # 5 épocas
    running_loss = 0.0
    for i, data in enumerate(train_loader, 0):
        inputs, labels = data

        optimizer.zero_grad()
        outputs = model_pytorch(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        running_loss += loss.item()
        if i % 100 == 99:  # print every 100 mini-batches
            print(f'Epoch {epoch + 1}, Batch {i + 1}: Loss {running_loss / 100:.3f}')
            running_loss = 0.0
end_time = time.time()
pytorch_time = end_time - start_time


print('Finished Training')

Epoch 1, Batch 100: Loss 0.973
Epoch 1, Batch 200: Loss 0.321
Epoch 1, Batch 300: Loss 0.220
Epoch 1, Batch 400: Loss 0.146
Epoch 1, Batch 500: Loss 0.126
Epoch 1, Batch 600: Loss 0.099
Epoch 1, Batch 700: Loss 0.098
Epoch 1, Batch 800: Loss 0.087
Epoch 1, Batch 900: Loss 0.078
Epoch 2, Batch 100: Loss 0.073
Epoch 2, Batch 200: Loss 0.059
Epoch 2, Batch 300: Loss 0.064
Epoch 2, Batch 400: Loss 0.067
Epoch 2, Batch 500: Loss 0.061
Epoch 2, Batch 600: Loss 0.057
Epoch 2, Batch 700: Loss 0.050
Epoch 2, Batch 800: Loss 0.054
Epoch 2, Batch 900: Loss 0.054
Epoch 3, Batch 100: Loss 0.044
Epoch 3, Batch 200: Loss 0.043
Epoch 3, Batch 300: Loss 0.052
Epoch 3, Batch 400: Loss 0.047
Epoch 3, Batch 500: Loss 0.036
Epoch 3, Batch 600: Loss 0.047
Epoch 3, Batch 700: Loss 0.035
Epoch 3, Batch 800: Loss 0.043
Epoch 3, Batch 900: Loss 0.040
Epoch 4, Batch 100: Loss 0.037
Epoch 4, Batch 200: Loss 0.032
Epoch 4, Batch 300: Loss 0.031
Epoch 4, Batch 400: Loss 0.037
Epoch 4, Batch 500: Loss 0.032
Epoch 4,

**RESULTADOS**

TEMPO E MÉTRICAS TENSORFLOW

In [ ]:
print(f"Tempo de treinamento Tensorflow: {keras_time:.2f} segundos")
# Avalia o modelo em acurácia
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print(f'\nTest accuracy: {test_acc}')

Tempo de treinamento Tensorflow: 357.11 segundos
313/313 - 4s - 14ms/step - accuracy: 0.9867 - loss: 0.0440

Test accuracy: 0.9866999983787537


In [ ]:
# Faz as predições
y_pred = np.argmax(model.predict(x_test), axis=-1)

# Avalia o modelo usando matriz de confusão e classification report
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step
Confusion Matrix:
[[ 969    0    2    0    0    2    4    3    0    0]
 [   0 1123    5    3    0    0    0    2    2    0]
 [   0    0 1022    2    0    0    0    7    1    0]
 [   0    0    1 1007    0    2    0    0    0    0]
 [   0    0    5    0  959    0    0    0    7   11]
 [   0    0    1    6    0  884    1    0    0    0]
 [   2    3    1    0    3    7  938    0    4    0]
 [   0    1    3    1    0    0    0 1018    0    5]
 [   0    0    1    5    0    2    0    3  962    1]
 [   1    0    1    4    0    4    0    3   11  985]]

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      0.99       980
           1       1.00      0.99      0.99      1135
           2       0.98      0.99      0.99      1032
           3       0.98      1.00      0.99      1010
           4       1.00      0.98      0.99       982
           5       0.98      0.99      0.99       892
       

TEMPO E MÉTRICAS PYTORCH

In [ ]:
print(f"Tempo de treinamento PyTorch: {pytorch_time:.2f} segundos")

# Testar o modelo
all_labels = []
all_predictions = []
correct = 0
total = 0
with torch.no_grad():
    for data in test_loader:
        images, labels = data
        outputs = model_pytorch(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
        all_labels.extend(labels.cpu().numpy())
        all_predictions.extend(predicted.cpu().numpy())

print(f'Accuracy of the network on the 10000 test images: {100 * correct / total:.2f}%')

Tempo de treinamento PyTorch: 248.17 segundos
Accuracy of the network on the 10000 test images: 98.93%


In [ ]:
print("Confusion Matrix:")
print(confusion_matrix(all_labels, all_predictions))

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions))


Confusion Matrix:
[[ 978    0    0    0    0    0    1    1    0    0]
 [   0 1133    1    1    0    0    0    0    0    0]
 [   1    1 1025    0    2    0    0    3    0    0]
 [   0    0    2 1004    0    2    0    1    1    0]
 [   0    0    0    0  976    0    0    0    0    6]
 [   2    0    0   14    0  867    2    1    2    4]
 [   3    2    0    1    3    2  946    0    1    0]
 [   0    3    4    0    1    0    0 1016    1    3]
 [   3    1    4    2    1    0    0    2  956    5]
 [   1    1    1    0    3    3    0    8    0  992]]

Classification Report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      1.00      1135
           2       0.99      0.99      0.99      1032
           3       0.98      0.99      0.99      1010
           4       0.99      0.99      0.99       982
           5       0.99      0.97      0.98       892
           6       1.00      0.99      0.99     

**ANÁLISE COMPARATIVA**

As duas soluções trouxeram resultados divergentes em tempo e muito semelhantes em métricas

* Tempo de treinamento Tensorflow: 357.11 segundos
* Tempo de treinamento PyTorch: 248.17 segundos

* Acurácia Tensorfow: 98.67%
* Acurácia PyTorch: 98.93%

A matriz de confusão com as métricas precision f1-score e recall foram semelhantes em todas as classes

**CONCLUSÃO**

1. Tempo de Treinamento

  * O PyTorch treinou o modelo significativamente mais rápido (cerca de 30% mais rápido) em comparação ao TensorFlow. Se o tempo de treinamento é um fator crítico, PyTorch pode ser uma melhor escolha.

2. Desempenho do Modelo

  * Ambas as implementações fornecem resultados de acurácia e métricas muito semelhantes, com uma ligeira vantagem para o PyTorch. Essa diferença pode ser considerada insignificante para a maioria das aplicações, indicando que ambos os frameworks são capazes de treinar modelos altamente eficazes.

3. Facilidade de Uso e Flexibilidade:

  * TensorFlow, com Keras, oferece uma interface de alto nível que pode ser mais intuitiva e rápida para o desenvolvimento, enquanto PyTorch é frequentemente preferido para experimentação e pesquisa devido à sua flexibilidade e integração com o Python nativo.

  Ambos os frameworks são altamente capazes de produzir modelos precisos e eficazes. A escolha entre TensorFlow e PyTorch pode depender mais de fatores como tempo de treinamento, preferências de desenvolvimento, e requisitos de produção. Para tarefas que exigem treinamento rápido, PyTorch pode ser preferível, enquanto TensorFlow pode ser a escolha para sistemas de produção robustos com suporte para deployment escalável.